src/qcodes/extensions/infer.py

```diff
@@ -1,16 +1,18 @@
 from __future__ import annotations
 
 from collections.abc import Sequence
-from typing import TYPE_CHECKING, ClassVar
+from typing import TYPE_CHECKING, ClassVar, TypeVar, cast
 
 from qcodes.instrument import Instrument, InstrumentBase, InstrumentModule
-from qcodes.parameters import DelegateParameter, Parameter
+from qcodes.parameters import DelegateParameter, Parameter, ParameterBase
 
 if TYPE_CHECKING:
     from collections.abc import Iterable
 
 DOES_NOT_EXIST = "Does not exist"
 
+C = TypeVar("C", bound=ParameterBase)
```

Comment:
Hi Samantha, methods you added here are quite similar to the one that you suggested me to add about one month ago. The difference is that I need a method to find specific type of instrument from delegation chain. Could you make `C` bounded to `Union[ParameterBase, InstrumentBase]` so new methods can also be used to find instruments?

Reply:
If we do that, then I'd like to add the instrument lookup versions too. 

Alternatively, we could add the instrument lookup methods in a separate PR and update the TypeVar then. For expediency, I prefer this second option.

Reply:
Ok, I'll create a PR to update TypeVar.